internal/postrs/api.go

```diff
@@ -0,0 +1,122 @@
+package postrs
+
+import (
+	"errors"
+	"fmt"
+)
+
+func OpenCLProviders() ([]Provider, error) {
+	return cGetProviders()
+}
+
+func CPUProviderID() uint {
+	return cCPUProviderID()
+}
+
+// ScryptPositionsResult is the result of a ScryptPositions call.
+type ScryptPositionsResult struct {
+	Output      []byte  // The output of the scrypt computation.
+	IdxSolution *uint64 // The index of a solution to the proof of work (if checked for).
+}
+
+type option struct {
+	providerID *uint
+
+	commitment    []byte
+	startPosition uint64
+	endPosition   uint64
+	n             uint32
+	vrfDifficulty []byte
+}
+
+func (o *option) validate() error {
+	if o.providerID == nil {
+		return errors.New("`providerID` is required")
+	}
+
+	if o.commitment == nil {
+		return errors.New("`commitment` is required")
+	}
+
+	if o.startPosition > o.endPosition {
+		return fmt.Errorf("invalid `startPosition` and `endPosition`; expected: start <= end, given: %v > %v", o.startPosition, o.endPosition)
+	}
+
+	if o.n > 0 && o.n&(o.n-1) != 0 {
+		return fmt.Errorf("invalid `n`; expected: power of 2, given: %v", o.n)
+	}
+
+	if o.vrfDifficulty == nil {
+		return errors.New("`vrfDifficulty` is required")
+	}
+
+	return nil
+}
+
+type OptionFunc func(*option) error
+
+// WithProviderID sets the ID of the openCL provider to use.
+func WithProviderID(id uint) OptionFunc {
+	return func(opts *option) error {
+		opts.providerID = new(uint)
+		*opts.providerID = id
+		return nil
+	}
+}
+
+func WithCommitment(commitment []byte) OptionFunc {
+	return func(opts *option) error {
+		if len(commitment) != 32 {
+			return fmt.Errorf("invalid `commitment` length; expected: 32, given: %v", len(commitment))
+		}
+
+		opts.commitment = commitment
+		return nil
+	}
+}
+
+func WithStartAndEndPosition(start, end uint64) OptionFunc {
+	return func(opts *option) error {
+		opts.startPosition = start
+		opts.endPosition = end
+		return nil
+	}
+}
+
+func WithScryptN(n uint32) OptionFunc {
+	return func(opts *option) error {
+		opts.n = n
+		return nil
+	}
+}
+
+func WithVRFDifficulty(difficulty []byte) OptionFunc {
+	return func(opts *option) error {
+		if len(difficulty) != 32 {
```

Comment:
The check wouldn't be necessary if the segment was [32]byte.

Reply:
Yes, but then the caller would need to do the check instead unless we can somehow ensure that difficulty is `[32]byte` everywhere. Generally I prefer `[]byte` over `[32]byte` because converting from `[32]byte` to `[]byte` is cheap:

```golang
var param [32]byte
WithVRFDifficulty(param[:]) // [:] does not copy the underlying array
```

but the other way round is more involved and expensive:

```golang
param := make([]byte, 32)
var difficulty [32]byte
if len(param) != 32 {
    // in case len(param) is not guaranteed to be 32
    return fmt.Errorf("len(param) not 32")
}

copy(difficulty[:], param) // copy necessary
WithVrfDifficulty(difficulty)
```

The functions calculating the difficulty also return `[]byte` and changing them to return `[32]byte` just shifts the conversation code into them.

Reply:
What other difficulties do you have in mind? If there is another kind of difficulty somewhere else, then it might have a different length or even type. I don't see problem here.

IMO an API accepting an array as argument is self descriptive and clear - user doesn't need to figure out what the difficulty is and it doesn't need to be documented. The runtime check is moved to compile time. An invalid code won't compile instead of failing in runtime.

Regarding conversion slice ->array, user can just use an array and avoid coversion.

Reply:
There are no other difficulties, the difficulty is calculated for the size of the PoST during runtime here:

https://github.com/spacemeshos/post/blob/9393f007d9475c860a8cedf13d20ea4fd41caad5/shared/shared.go#L51-L58

`(*big.Int).FillBytes` takes a `[]byte` as parameter, we would need to then copy the result into a new variable of type `[32]bytes` and return this (= another copy unless we return it as a `*[32]byte`). `[]byte` are passed via reference by default, `[32]bytes` are passed via value unless we use `*[32]byte`. 

The check here is just as a safety precaution (and cheap -> `len([]byte)` is `O(1)`). In our code it will never be a different length, but to enforce that during compile time makes the code less efficient, harder to read and less convenient to use imo.

Reply:
See also https://go.dev/doc/effective_go#arrays and https://go.dev/blog/slices-intro

Reply:
I'm almost sure it's optimized nicely. Nevertheless, who cares about efficiency here? It's not a hot path.

IMHO, the code using array is actually more readable because the constraints are clearly presented in the API, which is more important. In other words, the API is easier to use.

Reply:
I'm not agreeing with it makes the API easier to use, it might be idiomatic in Rust but in Go arrays are rarely used and imo this is not one of the places where it makes sense to use it.

Having the `WithVRFDifficulty` take a `[32]byte` instead of a `[]byte` requires the following changes:

Change this function https://github.com/spacemeshos/post/blob/9393f007d9475c860a8cedf13d20ea4fd41caad5/shared/shared.go#L51-L58

to 

```golang
func PowDifficulty(numLabels uint64) [32]byte { 
 	x := new(big.Int).Lsh(big.NewInt(1), 256) 
 	x.Div(x, big.NewInt(int64(numLabels))) 
 	x.Lsh(x, 3) // reduce difficulty by a factor of 8 
  
 	difficulty := make([]byte, 32)
 	x.FillBytes(difficulty)
	var result [32]byte
	copy(result[:], difficulty)
	return result
 }
```

and then where it is actually used:

https://github.com/spacemeshos/post/blob/9393f007d9475c860a8cedf13d20ea4fd41caad5/internal/postrs/initializer.go#L95-L100

convert it back to a slice?

```golang
	difficulty := opt.vrfDifficulty[:]
	cDifficulty := C.CBytes(difficulty)
	defer C.free(cDifficulty)
	init := C.new_initializer(cProviderId, cN, (*C.uchar)(cCommitment), (*C.uchar)(cDifficulty))
	if init == nil {
		return nil, nil, ErrInvalidProviderID
	}
```

> Nevertheless, who cares about efficiency here? It's not a hot path.

To some degree it is. To verify VRF nonces of incoming ATXs I have to initialize the one label that is referenced by the nonce and pass in the required difficulty (see `verifying.VerifyVRFNonce`). Depending on how many smeshers are participating in the network this might be called quite often - also consider that if a smesher wants to initialize a large PoST they probably have to do this over multiple epochs and keep updating publishing new ones 🙂